### PR TITLE
Lint for for loops over indexes instead of items

### PR DESCRIPTION
**Pull request by birkenfeld**

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -109,15 +109,25 @@ pub fn walk_ptrs_ty<'t>(ty: ty::Ty<'t>) -> ty::Ty<'t> {
 ///         }
 ///     }
 #[macro_export]
-macro_rules! if_let_chain {
```
Comment (Manishearth) [inline src/utils.rs]: Why rename the macro? It's a chain of `if let`s, not `if`s, the two aren't interchangeable.

Comment (Manishearth) [inline src/utils.rs]: Oh nevermind I see what you did here, this is awesome!

I still think the macro should be called `if_let_chain` though.

File: src/loops.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                if let ExprRange(_, _) = arg.node {
+                    let mut visitor = VarVisitor { cx: cx, var: ident.node.name,
+                                                   indexed: HashSet::new(), nonindex: false };
+                    walk_expr(&mut visitor, body);
+                    // linting condition: we only indexed one variable
+                    if visitor.indexed.len() == 1 {
+                        let indexed = visitor.indexed.into_iter().next().unwrap();
+                        if visitor.nonindex {
+                            span_lint(cx, NEEDLESS_RANGE_LOOP, expr.span, &format!(
+                                "the loop variable `{}` is used to index `{}`. Consider using \
```
Comment (Manishearth) [inline src/loops.rs]: nit: "only used"

File: src/loops.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                // we are referencing our variable! now check if it's as an index
+                if_chain! {
+                    [
+                        let Some(parexpr) = get_parent_expr(self.cx, expr),
+                        let ExprIndex(ref seqexpr, _) = parexpr.node,
+                        let ExprPath(None, ref seqvar) = seqexpr.node,
+                        seqvar.segments.len() == 1
+                    ], {
+                        self.indexed.insert(seqvar.segments[0].identifier.name);
+                        return;  // no need to walk further
```
Comment (Manishearth) [inline src/loops.rs]: This won't stop the visitor from walking, fwiw

**Comment by Manishearth** [comment]
LGTM (sorry about the erroneous comments -- sleepy and didn't read through the code -- deleted now).

Bikeshed about the macro name? It's main purpose is to nest `if let`s, so I prefer `if_let_chain`, but don't feel too strongly about it.

**Comment by Havvy** [comment]
I was just being lazy coming up with a name when I wrote my example, so I went with the first thing I could think of. Given a bit more though on humane naming, I'd say `if_all` would be a better name.

**Comment by Manishearth** [comment]
if_all is even more confusion to me :p

**Comment by Manishearth** [comment]
(The name should at least have chain in it)

**Comment by birkenfeld** [comment]
Ok, there we go :)

**Pull request merged**